.pre-commit-config.yaml

```diff
@@ -15,17 +15,18 @@ repos:
           - flake8-docstrings==1.3.1
           - pydocstyle==4.0.0
         files: ^(homeassistant|script|tests)/.+\.py$
-# Using a local "system" mypy instead of the mypy hook, because its
-# results depend on what is installed. And the mypy hook runs in a
+# Using a locally set up mypy instead of the "real" mypy hook, because
+# mypy's results depend on what is installed. The "real" one runs in a
 # virtualenv of its own, meaning we'd need to install and maintain
-# another set of our dependencies there... no. Use the "system" one
-# and reuse the environment that is set up anyway already instead.
+# another set of our dependencies there... no. Use a wrapper script
+# that launches the one installed in the currently active environment
+# that is set up as desired instead.
 -   repo: local
     hooks:
     -   id: mypy
-        name: mypy
-        entry: mypy
-        language: system
+        name: mypy (if available)
+        entry: script/mypy_wrapper.py
```

Comment:
thinking now, could we run mypy with `--ignore-missing-imports` instead of making it this complicated?

Reply:
We already do, through config in setup.cfg.

Reply:
^we do the ignore missing imports part of that in the config, that is.

Running inside our venv is what we would like, sure -- if we can do and ensure that, then this stuff can be fixed. But I don't think it's sanely/bulletproof doable. Simply hardwiring knowledge what/where the venv actually is on end user systems would be required, and that might complicate or even practically possibly rule out use of e.g. pyenv-virtualenv. Not being able to use that would suck.

Reply:
A further complication for the above, if you mean writing a script that activates the venv and then runs mypy in it, would be Windows. pre-commit shell script would be a no go for that reason I guess.

_If_ we could assume the venv is always in `venv`, then I suppose invoking `venv/bin/mypy` with a `language: script` entry would make that stuff happen, but see my previous comment for drawbacks with that assumption. I dunno if that would work on Windows either, or if the script would have a different name and need to be called some other way there.

Reply:
Well I actually meant, we could run mypy inside the pre-commit venv with the `--ignore-missing-imports` option, and we would already have a good, but not perfect, type check.

Reply:
A problem with that is in such a setup, we may need for example `type: ignore`s etc workarounds that are not needed in "complete  with deps env" checks or vice versa, and they may get flagged as errors in the other. We've already been there at least with ruamel.yaml in differently configured envs, and there's more I'm sure. It's probably possible to find a set of settings that work in both, but I believe it would be the "worst" denominator of all of them, and I'm not at all convinced that spending time on that is worth it.

My current feeling is that we'd be best off just removing mypy from pre-commit at least for now.

Reply:
Yeah, I guess you're right. It sucks because it has helped me catch things for sure, since I work mainly in core.

Reply:
Actually, there is one more in-between option, let me submit another PR for that.

Reply:
#28463